[RFC] Break up screen grid into resizable window grids

coditva opened the pull request:
Neovim has one global grid for the screen in the form of `ScreenLines`, `ScreenAttrs` etc.
Although it is very rough around the edges, it breaks up this global grid into `ScreenGrid`s which are unique for every window. It also keeps a `default_grid` which is accessible globally and takes care of `vseps`, `statuslines` etc.

**Update:** The API now allows UIs to set grid sizes for each grid individually. UIs receive grid and position updates for each window separately. It allows them to use different font sizes for each window if they choose so, implement scrollbars in their own toolkit, etc. while `nvim` handles the wincmds. More info: [blog post](https://utkarshme.github.io/blog/window-grid-updates-in-neovim).

**Next up:** The next level of API (soon to follow) is expected to delegate the handling of wincmds to the UIs.

This work builds on top of previous work by @bfredl in #6619 and #8221. Ref #8320

bfredl committed:
multigrid: introduce grid abstraction

bfredl committed:
multigrid: add multigrid support to test infrastructure

make Screen explicitly tied to its session

coditva committed:
multigrid: Change screen_* functions to grid_* functions

bfredl commented on src/nvim/screen.c:
```diff
@@ -5017,9 +5017,9 @@ win_redr_custom (
     use_sandbox = was_set_insecurely((char_u *)"tabline", 0);
   } else {
-    row = wp->w_height;
+    row = wp->w_grid.Rows + wp->w_grid.OffsetRow;
```
it is probably better to use the original `wp->w_winrow+wp->w_height` here. Because when we will allow UI:s to change the "inner" grid size, the statusline shouldn't be affected. So in general (i e except when we have a reason to do otherwise), code that draws on global grid should use `wp->w_*` varables, code that draws on local grid should use `wp->w_grid.*` variables.

bfredl commented:
There is a slight issue: when a UI attaches, it will receive a `grid_resize` for the global grid, but not necessarily for the local grids, because the later will only be sent if the size actually changed. (And I think nvim's internal size default is the same as python-gui's default) We can fix this by have `screen_resize` set a flag, so that during the next `update_screen`, when `window_grid_alloc` (note: `window_` should be `win_` here) is called, the `gird_resize` event is sent even if the grid wasn't reallocated.

bfredl commented:
In general I think the approach for `ui_line` is the right one, you could do more or less the same for `ui_cursor_goto`, the two imporant cases are `setcursor()` which should use the window grid variables, and `screen_puts_line_start/flush` where we can add `put_dirty_grid` (which the start method should set).

bfredl commented:
Now, the time to the first evaluation is running a bit short to introduce API to change grid size. But here is a cheap trick we could use to at least fix/verify that the new implemenation can use "inner" grid size independent of split allocation. On the condition that kUIMultigrid is active, change

    grid_alloc(&wp->w_grid, wp->w_height, wp->w_width, doclear);

to something like
    
    grid_alloc(&wp->w_grid, (int)(1.5*wp->w_height), (int)(1.5*wp->w_width), doclear);

(but probably precomputing the inner size to a variable to handle the surrounding check also). This will make us discover most cases where w_height is used where w_grid->Rows should be used or vice versa.

coditva committed:
multigrid: Allow UIs to set grid size different from window size

bfredl commented on src/nvim/screen.c:
```diff
@@ -5823,8 +5823,10 @@ int screen_valid(int doclear)
   if (wp->w_grid.ScreenLines == NULL
-      || wp->w_grid.internal_Rows != wp->w_height
-      || wp->w_grid.internal_Columns != wp->w_width) {
+      // TODO(utkarshme): these lines create a redraw very frequently. find the problem
```
The problem is that you are comparing the "outer" size, which is irrelevant here. You want to calculate `1.5 * wp->w_height` to a local variable, and then compare that to `w_grid.Rows`.

bfredl replied:
Or maybe we should just delete the 1.5 thing, now that external size control is almost working.

bfredl commented on src/nvim/ui.c:
```diff
@@ -448,3 +448,15 @@ Array ui_array(void)
+    abort();
+    return;
+  }
+  grid_alloc(grid, height, width, true);
```
We need to trigger a redraw also. If the grid belongs to a window, can use `redraw_win_later` (probably SOME_VALID, but not sure).

Actually we probably shouldn't reallocate here, redraw will always reallocate (and call `grid_resize` back)  if necessary anyway. Instead we shoud store the intended size so that futher redraws will respect it (and not overwrite it with the "outer" size), maybe in `internal_Rows`.

bfredl commented on src/nvim/types.h:
```diff
@@ -40,8 +40,9 @@ typedef struct {
-  int internal_Rows;
-  int internal_Columns;
+  // the size expected to be allocated to the internal grid
+  int InternalRows;
```
newly introduced variables should use `lower_case` (previously global varables moved to a struct is more of a border case, but that is something that we can deal with later)

coditva committed:
multigrid: Add win_position event

Throttle win_position events

coditva committed:
multigrid: Put everything on default_grid if not ext_multigrid

coditva committed:
multigrid: Add multigrid documentation

breja commented:
Super excited for this PR! Many thanks to @UtkarshMe and @bfredl for the work on this!

I know this is still WIP but I was really curious to mess around with this. I apologize ahead of time if I'm jumping in too early, don't know what I'm talking about, or just causing noise.

I had some observations and was wondering if I could get some clarity. I put together a simple test app using this PR to print out the new/multi grid + win_position events when sending some window commands.

I noticed that when creating a new window (e.g. `:vsplit` or `:split`) the `win_position` grid id of the new window is `0`. I would have expected the grid ids to match with the ones being sent by `grid_resize`.
- e.g. in the output below, after calling `:vsplit` I would have expected the first/current window to be grid id `2` and the new split window to be grid id `3` (because the `grid_resize` event sends `2` and `3` respectively)

Is this intentional or do I misunderstand? Does grid `0` have any meaning? I believe grid `1` is the global, grid `2` is the first window, grid `3` second window, and so on.

Interestingly enough, when doing a resize event, `win_position` sends grid ids that match with the ids sent by `grid_resize`

Is your experience the same?

### test app log output

```
--> nvim_ui_attach(80, 40, { ext_multigrid: true })

grid_resize(grid: 1, width: 80, height: 40)
win_position(win: 1000, grid: 2, top: 0, left: 0, width: 80, height: 38)
grid_clear(grid: 1)
grid_resize(grid: 2, width: 80, height: 38)
grid_resize(grid: 2, width: 80, height: 38)

--> :vsplit
win_position(win: 1001, grid: 0, top: 0, left: 0, width: 80, height: 38)
win_position(win: 1001, grid: 0, top: 0, left: 0, width: 40, height: 38)
win_position(win: 1000, grid: 2, top: 0, left: 0, width: 39, height: 38)
win_position(win: 1000, grid: 2, top: 0, left: 41, width: 39, height: 38)
win_position(win: 1001, grid: 0, top: 0, left: 0, width: 40, height: 38)
grid_resize(grid: 3, width: 40, height: 38)
grid_resize(grid: 2, width: 39, height: 38)

--> :split
win_position(win: 1002, grid: 0, top: 0, left: 0, width: 40, height: 0)
win_position(win: 1002, grid: 0, top: 0, left: 0, width: 40, height: 19)
win_position(win: 1001, grid: 3, top: 0, left: 0, width: 40, height: 18)
win_position(win: 1001, grid: 3, top: 20, left: 0, width: 40, height: 18)
win_position(win: 1002, grid: 0, top: 0, left: 0, width: 40, height: 19)
grid_resize(grid: 4, width: 40, height: 19)
grid_resize(grid: 3, width: 40, height: 18)

--> :resize -2
win_position(win: 1002, grid: 4, top: 0, left: 0, width: 40, height: 17)
win_position(win: 1001, grid: 3, top: 20, left: 0, width: 40, height: 20)
grid_resize(grid: 4, width: 40, height: 17)
grid_resize(grid: 3, width: 40, height: 20)
```

**repo**: https://github.com/breja/nvwin

Thanks again for your time

UtkarshMe commented:
@breja, it's very motivating to see people take interest. Thank you!! :heart:  Any effort put by you is much appreciated.

@bfredl On it! :+1:

breja commented:
Thanks for the quick reply and fix! I am seeing the correct grid ids now in `win_position`

I played around some more and I noticed that when closing the first of two windows, the window positions do not seem to update in `win_position`

For example:
- create one vertical or horizontal split - `:vsplit` or `:split`
- move focus back to first/original window - `<c-w>h` or `<c-w>k`
- close window - `:q` or `<c-w>c`
- win_position `top` or `left` of remaining window does not update
  - since this is the last and only window now, I would expect both `top` and `left` to be `0`

Another observation - I noticed that `grid_cursor_goto` does not always update and/or send the correct grid_id when changing focus from one window to another (you can see an example of this in the output below). I did not collect enough data to make a consistent reproducible test case - will report back on this.

Thanks again for the time and effort!

## output

lines of interest start with `?? >>>`

- vertical
```
--> nvim_ui_attach(80, 40, {"ext_multigrid":true})

--REDRAW--
win_position(win: 1000, grid: 2, top: 0, left: 0, width: 80, height: 38)
grid_cursor_goto(id: 2, row: 0, col: 0)
--REDRAW--
--> nvim_command("vsplit")
--REDRAW--
win_position(win: 1001, grid: 3, top: 0, left: 0, width: 80, height: 38)
win_position(win: 1001, grid: 3, top: 0, left: 0, width: 40, height: 38)
win_position(win: 1000, grid: 2, top: 0, left: 0, width: 39, height: 38)
win_position(win: 1000, grid: 2, top: 0, left: 41, width: 39, height: 38)
win_position(win: 1001, grid: 3, top: 0, left: 0, width: 40, height: 38)
grid_cursor_goto(id: 3, row: 0, col: 0)
--> nvim_input("<c-w>h")
--REDRAW--

?? >>> grid_cursor_goto(id: 3, row: 0, col: 0)

--REDRAW--
grid_cursor_goto(id: 3, row: 0, col: 0)
--> nvim_command("q")
--REDRAW--

?? >>> win_position(win: 1000, grid: 2, top: 0, left: 41, width: 80, height: 38)

destroy(grid: 3)
grid_cursor_goto(id: 2, row: 0, col: 0)
```

- horizontal
```
--> nvim_ui_attach(80, 40, {"ext_multigrid":true})

--REDRAW--
win_position(win: 1000, grid: 2, top: 0, left: 0, width: 80, height: 38)
grid_cursor_goto(id: 2, row: 0, col: 0)
--REDRAW--
--> nvim_command("split")
--REDRAW--
win_position(win: 1001, grid: 3, top: 0, left: 0, width: 80, height: 0)
win_position(win: 1001, grid: 3, top: 0, left: 0, width: 80, height: 19)
win_position(win: 1000, grid: 2, top: 0, left: 0, width: 80, height: 18)
win_position(win: 1000, grid: 2, top: 20, left: 0, width: 80, height: 18)
win_position(win: 1001, grid: 3, top: 0, left: 0, width: 80, height: 19)
grid_cursor_goto(id: 3, row: 0, col: 0)
--> nvim_input("<c-w>k")
--REDRAW--

?? >>> grid_cursor_goto(id: 3, row: 0, col: 0)

--REDRAW--
grid_cursor_goto(id: 3, row: 0, col: 0)
--> nvim_command("q")
--REDRAW--

?? >>> win_position(win: 1000, grid: 2, top: 20, left: 0, width: 80, height: 38)

destroy(grid: 3)
grid_cursor_goto(id: 2, row: 0, col: 0)
```

UtkarshMe commented:
Hi @breja, thanks for testing!

> `grid_cursor_goto` does not always update and/or send the correct `grid_id`

In the `:vsplit` case, the new split (i.e. `grid 3`) is created on the left and we switch to it. You will see that `grid 2` has `left:41` in `win_position`. You should be pressing `<C-W> l` to see the `grid_cursor_goto(id:2 ...)`. Likewise for the `:split` case when window is created on the top.


> the window positions do not seem to update in `win_position`

You are correct about this. I'll push a fix as soon as possible. Thank you for the time and input! :smile:

UtkarshMe commented:
@bfredl Should we send only one latest `win_position` event for a specific grid or is the current scenario fine?

bfredl commented:
@UtkarshMe If it wouldn't be too difficult, ideally maximum one `win_position` per window for a elementary wincmd like `split`.

UtkarshMe commented on src/nvim/api/ui.c:
```diff
@@ -275,18 +275,27 @@ static void push_call(UI *ui, const char *name, Array args)
+      int did = 0;
+      // the adjacent calls are bundled together into one call so we must
+      // go through all the calls to find and update the relevant calls.
+      for (int i = 1; i < kv_size(temp_call); i++) {
```
@bfredl This achieves what has to be done, but looks a little too complicated(?) Any suggestion about how we can make this not look ugly? :sweat_smile: Or is it fine? :thinking:

bfredl replied:
Where are these events generated in the first place in `window.c`? How tricky would it be to throttle them there?

breja commented:
Thanks again for the fix!

I noticed that sometimes windows do not update positions after another window is closed. I'm not sure if it's just me or not.

For example:
- `vsplit`
- `split`
- `ctrl-w h` (to first/left window)
- `ctrl-w c`

(Bottom right window `left` position does not update to `0`)

layout:
```
|   |   |
|   |---|
|   |   |
```

demo:
![extwin3](https://user-images.githubusercontent.com/2581815/43235297-37aa09ba-904e-11e8-8c7e-14af1f6cea63.gif)

logs (see line of interest `-->`):
```
grid_resize(grid: 1, width: 98, height: 26)
win_position(win: 1000, grid: 2, top: 0, left: 0, width: 98, height: 25)
grid_clear(id: 1)
grid_resize(grid: 2, width: 98, height: 25)
grid_resize(grid: 2, width: 98, height: 25)
grid_cursor_goto(id: 2, row: 0, col: 0)
grid_cursor_goto(id: 1, row: 25, col: 25)
grid_cursor_goto(id: 1, row: 25, col: 38)
grid_cursor_goto(id: 2, row: 0, col: 0)

vsplit

win_position(win: 1001, grid: 3, top: 0, left: 49, width: 49, height: 25)
win_position(win: 1000, grid: 2, top: 0, left: 0, width: 48, height: 25)
grid_resize(grid: 2, width: 48, height: 25)
grid_resize(grid: 3, width: 49, height: 25)
grid_cursor_goto(id: 3, row: -1, col: -1)

split

win_position(win: 1002, grid: 4, top: 13, left: 49, width: 49, height: 12)
win_position(win: 1001, grid: 3, top: 0, left: 49, width: 49, height: 12)
grid_resize(grid: 3, width: 49, height: 12)
grid_resize(grid: 4, width: 49, height: 12)
grid_cursor_goto(id: 1, row: 12, col: 72)
grid_cursor_goto(id: 4, row: -1, col: -1)

ctrl-w h

grid_cursor_goto(id: 2, row: 0, col: 0)

ctrl-w c

win_position(win: 1001, grid: 3, top: 0, left: 49, width: 98, height: 12)
---> (left should be 0?) --> win_position(win: 1002, grid: 4, top: 13, left: 49, width: 98, height: 12)
grid_destroy(id: 2)
win_position(win: 1001, grid: 3, top: 0, left: 0, width: 98, height: 12)
grid_resize(grid: 3, width: 98, height: 12)
grid_resize(grid: 4, width: 98, height: 12)
grid_cursor_goto(id: 1, row: 12, col: 23)
grid_cursor_goto(id: 3, row: 0, col: 0)
```

bfredl commented:
@UtkarshMe could the issue @breja sees be related to the way we try to implement throttling of `win_position`?

 As I said before, it would be preferable if wincmd code in window.c could just emit these events once per user-level wincmd,  but failing that I think throttling could be implemented in a cleaner way than messing with the msgpack structs: add a bool to `win_T` to mark a window as "dirty" whenever a win_position now is sent, then in `ui_flush` add a call to a function in window.c that goes through all windows and send one `win_position` for each window in the current tab that was marked as dirty. (this will be consistent with how the throttle ext_cmdline events, for instance).

UtkarshMe commented:
@bfredl, there does not seem to be any issue with the throttling (the error is reproducible before and after the throttling commits). But I was able to find the bug (not enough `win_position`s sent). Minimizing the number of `ui_win_position` will take refactor of `window.c/frame_comp_pos` which I find unrelated to this PR. I think we might be better off adding the "dirty" bool as you suggested.

bfredl commented:
> I think we might be better off adding the "dirty" bool as you suggested.

Sounds reasonable.

bfredl commented on src/nvim/edit.c:
```diff
@@ -1488,8 +1488,8 @@ void edit_putchar(int c, int highlight)
         if (fix_col != pc_col) {
-          screen_putchar(' ', pc_row, fix_col, attr);
-          --curwin->w_wcol;
+          grid_putchar(&default_grid, ' ', pc_row, fix_col, attr);
```
I think we can use `wp->w_grid` here (and all other changed calls in this file), and remove `w_winrow`/`w_wincol`) and the new fallback logic should handle the not-multigrid case.

UtkarshMe replied:
I'm not sure if I should be removing `w_winrow` from this: https://github.com/UtkarshMe/neovim/blob/6313b02704e7ac357911e169c71881f5b43fcd9d/src/nvim/edit.c#L559
`curwin->w_wrow` is relative to the window i.e. `w_winrow`. I don't know what this line does.

Also, I should be changing `w_height`/`w_width` to `w_grid.Rows`/`w_grid.Columns` too, right?

bfredl commented on src/nvim/edit.c:
```diff
@@ -1488,8 +1488,8 @@ void edit_putchar(int c, int highlight)
-          screen_putchar(' ', pc_row, fix_col, attr);
-          --curwin->w_wcol;
+          grid_putchar(&default_grid, ' ', pc_row, fix_col, attr);
+          (curwin->w_wcol)--;
```
parenthesis not needed.

UtkarshMe commented:
At the end of the GSoC coding period, the state of this PR was [this](https://github.com/UtkarshMe/neovim/tree/grid-split-gsoc-end).

bfredl commented on runtime/doc/ui.txt:
```diff
@@ -30,7 +30,8 @@ a dictionary with these (optional) keys:
 	`ext_newgrid`		Use new revision of the grid events. |ui-newgrid|
-	`ext_hlstate`		Use detailed highlight state. |ui-hlstate|
+	`ext_multigrid`		Use per-window grid based events. |ui-multigrid|
+	`ext_hlstate`		Use detailed highlight state.  |ui-hlstate|
```
extra whitespace.

bfredl commented on src/nvim/api/ui.c:
```diff
@@ -270,6 +292,10 @@ static void remote_ui_grid_resize(UI *ui, Integer grid,
   Array args = ARRAY_DICT_INIT;
   if (ui->ui_ext[kUINewgrid]) {
     ADD(args, INTEGER_OBJ(grid));
+  } else if (grid != 1) {
```
do we still need this special case? I would think screen.c by itself will/should avoid inner resizes if multigrid is not active. (But If we still do, the condition should be `!ext_multigrid` and not only `!ext_newgrid` )

UtkarshMe replied:
Also, in `doc/ui.txt`, we have `grid_resize` under `ext_newgrid`. If `ext_newgrid` will be set when `ext_multigrid` is set so does that not cover it?

UtkarshMe replied:
Yes, I got that. What I meant was that `grid_resize` is listed in `ext_newgrid` functions in `ui.txt`. So, I was unsure if the check should be `ext_newgrid` or `ext_multigrid`. You said it should be `!ext_multigrid` and _not only_ `!ext_newgrid` hence the confusion. Sorry... :sweat:

bfredl replied:
> You said it should be !ext_multigrid and not only !ext_newgrid hence the confusion

That would be if we would keep the check (which we should not), sorry for the confusion.

bfredl commented on src/nvim/edit.c:
```diff
@@ -1514,12 +1514,14 @@ void edit_putchar(int c, int highlight)
-    else
-      screen_puts(pc_bytes, pc_row - msg_scrolled, pc_col, pc_attr);
+    if (pc_status == PC_STATUS_RIGHT) {
+      (curwin->w_wcol)++;
```
parens not needed.

bfredl commented on src/nvim/move.c:
```diff
@@ -134,7 +134,11 @@ void update_topline(void)
   // the cursor line.
-  if (!screen_valid(true) || curwin->w_height == 0) {
+  if (!screen_valid(true))
+    return;
```
what for this extra return?

UtkarshMe replied:
I think this slipped in during rebase :sweat_smile: .

bfredl commented on src/nvim/screen.c:
```diff
@@ -5833,7 +5881,57 @@ void check_for_delay(int check_msg_scroll)
+      || grid->Rows != rows
+      || grid->Columns != columns) {
+    grid_alloc(grid, rows, columns, doclear);
+    grid->was_resized = true;
```
can't this variable be local to this function?

bfredl commented on src/nvim/screen.c:
```diff
@@ -6017,14 +6047,63 @@ void screenalloc(bool doclear)
+
+void set_screengrid(ScreenGrid *grid)
+{
+  ScreenLines = grid->ScreenLines;
```
are we still using global ScreenLines somewhere? If not (or if it could easily be removed), this function can be eliminated.

UtkarshMe replied:
Not in `screen.c` but in other files. About 10-15 instances. Will change those.

bfredl commented on test/functional/ui/screen_multigrid_spec.lua:
```diff
@@ -0,0 +1,684 @@
+local helpers = require('test.functional.helpers')(after_each)
```
as the feature is named "multigrid", `multigrid_spec.lua` is a better name

coditva committed:
multigrid: Get rid of global ScreenLines and set_screengrid

coditva committed:
multigrid: Add ScreenGrid as a param to multibyte functions

coditva committed:
multigrid: Clean whole grid when it was just resized

bfredl committed:
multigrid: avoid allocation when not ext_multigrid.

Move grid specific functions from mbyte.c to screen.c

bfredl commented:
I took the liberty to try refactor the temporary line buffer logic [here](https://github.com/bfredl/neovim/tree/grid-split_alloc), to avoid any extra allocation when `ext_multigrid` is not active. All drawing is then by default done on a single grid only, just as before this PR. 

I think this is starting to get into reasonable shape for merging. Apart from some failing tests, some remaining points to consider:

- Default popupmenu rendering isn't really useful with ext_multigrid. We can workaround this for now by expecting/mandating multigrid UIs to use ext_popupmenu. We should make sure that the anchor position in `popupmenu_show` is relative to the current window's grid. Later on, the builtin popupmenu could be transmitted as a floating window for client that don't opt into `ext_popupmenu`.

- Long messages are shown by scrolling the default grid only, which is assumed to be "below" the window grids. We can provide UIs with a hint (UI event) when message scrolling starts (so that scrolled messages should overdraw windows) and another when scrolled messages are dismissed (so that window grids should be redrawn on top again). For UIs wanting more control, I hope to make ext_messages #7466 available soon.

-  This has some overlap/interactions with floating windows #6619. Implementation-wise, it is probably simplest to merge this PR first, and then I will refactor #6619 to work on top of this. Regarding the interface however, UIs that use ext_multigrid will be expected to also handle floating windows after #6619 is merged. This could be argued a breaking change (if a release is done in between), but I don't expect much trouble in practise, as floating windows should initially be considered a bit experimental, and only be used when users opt-in to them (by installing a float-specific plugin, or by configuring a generic plugin to use floats).

coditva committed:
multigrid: Fix sending window grid handle in ext_newline mode

coditva committed:
multigrid: Fix grid allocation misses

- Clear whole grid in one go.
- Fix wrongly sent "copy" flag.
- Add clear function comment.

bfredl commented on src/nvim/api/ui.c:
```diff
@@ -405,6 +405,8 @@ static void remote_ui_highlight_set(UI *ui, int id)
 static void remote_ui_grid_cursor_goto(UI *ui, Integer grid, Integer row,
                                        Integer col)
 {
+  grid = ui->ui_ext[kUIMultigrid] ? grid : DEFAULT_GRID_HANDLE;
```
It is better to do this in `ui.c`, where the offset is added (so the reference to the original grid doesn't make any sense anymore)

bfredl commented on src/nvim/screen.c:
```diff
@@ -6027,14 +6130,76 @@ void screenalloc(bool doclear)
+  new.Rows = rows;
+  new.Columns = columns;
+
+  memset(new.ScreenLines, ' ', ncells * sizeof(schar_T));
```
not correct, use `grid_clear_line` in the loop instead.

bfredl commented on src/nvim/screen.c:
```diff
@@ -665,19 +675,21 @@ static void win_update(win_T *wp)
 
   type = wp->w_redr_type;
 
+  win_grid_alloc(wp, true);
```
I'm not sure we want to default to clear window grids on window resize. It is better to define what happens with the remaining content on resize, to reduce redrawing when a separator is dragged, etc. It looks like the present code is prepared for preserving the bottom line (for gvim to behave like a terminal when "needed", but I don't think nvim currently uses it, it just refrains from resizing the grid in "press RETURN" state). But perhaps it would be simpler (for new GUIs) to preserve the topline as it is defined as the origin `(0,0)`.

bfredl commented on src/nvim/window.c:
```diff
@@ -5985,3 +6017,18 @@ void win_findbuf(typval_T *argvars, list_T *list)
+  FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
+    if(wp->w_pos_changed) {
+      // make sure grid is allocated before the ui gets any event for it
+      win_grid_alloc(wp, true);
```
doing allocation here seems like weird coupling. If we need to make sure the first grid state is sent in the same batch as the first win_position, it would better to delay the sending of `win_position` (by adding `wp->grid.ScreenLines != 0` or something at the if statement here), than to send win_position with a blank grid state in one batch and the real grid state in the next batch.

bfredl committed:
multigrid: use grid-based coordinates for ext_popupmenu

bfredl commented:
Here is a patch that makes ext_popupmenu transmitt the correct position: https://github.com/bfredl/neovim/commit/156bf89e212038b8e123aafdd0db598697347dd9
Per specification, appending new args to a method is a non-breaking change, as long as the existing behavior (ext_popupmenu but without ext_multigrid) is preserved by ignoring the arg.

UtkarshMe commented on test/functional/ui/screen.lua:
```diff
@@ -463,9 +479,9 @@ function Screen:_handle_mode_change(mode, idx)
 
 function Screen:_handle_set_scroll_region(top, bot, left, right)
-  self._scroll_region.top = top
+  self._scroll_region.top = top + 1
```
@bfredl This is causing many tests to fail... Why are we changing it? We would also need to make the change in `api/ui.c`, right?

bfredl commented:
> self._scroll_region.top = top + 1
> This is causing many tests to fail... Why are we changing it? We would also need to make the change in api/ui.c, right?

It could be a logical rebase conflict. If both modes of `screen_basic_test.lua` works without it, then removing it is correct...

UtkarshMe commented:
`screen_basic_spec` works flawlessly after removing, but some other tests start failing (for e.g. in `cmdline_highlight_spec.lua`).

bfredl commented:
I'll take a look later. I think the purpose is to make the internal representation to be the new one and translate from the old, rather than the other way around. (maybe #6619 contains the "correct" code, I don't remember).

coditva committed:
multigrid: Fix rebase errors in screen.c

coditva committed:
multigrid: Draw fold for the entire width of window grid

bfredl commented on src/nvim/move.c:
```diff
@@ -146,9 +146,14 @@ void update_topline(void)
   // If there is no valid screen and when the window height is zero just use
   // the cursor line.
-  if (!screen_valid(true) || curwin->w_height == 0) {
+  if (curwin->w_grid.Rows == 0) {
```
@UtkarshMe The patch that added the now failing `test_startup.vim` test added `screen_valid` here, why was it removed again?

UtkarshMe replied:
I remember changing it because some test was failing because of it. But now, reverting and re-running the tests, all seem to pass. Maybe there was some other problem which was causing it. Reverting back!

coditva committed:
multigrid: Fix lint errors

bfredl committed:
multigrid: don't clear window grids on resize

Instead define that the shared top-left part of the grid is preserved.

bfredl commented:
This patch should do what I thought for avoiding clearing on resize https://github.com/bfredl/neovim/commit/23f24337bdbce8b85e88ddca9b3ac4794d32c6cc . The change of `screen.lua` reflects the actual change of protocol.

Also, anywhere the old vim code clears the screen because "maybe it is faster", we should feel free to remove that clear (because the new grid_line allows clearing of large regions efficiently anyway). Strictly speaking the only time we need to clear a grid is when (1) the global grid is resized (because non-multigrid clients and TUI expect it) or (2) the user requested it, i e `<c-l>` or `:mode`.

bfredl commented on runtime/doc/ui.txt:
```diff
@@ -422,6 +428,35 @@ because 'winhighlight' was used. UI items will be transmitted, even if the
+`ext_newgrid` implicitly.
+
+The multigrid enables the UIs to manipulate window grids individually. The UIs
+receive a grid update for each new window. The UIs can set a grid size
```
more something like "updates on a separate grid for each new window"

bfredl commented on runtime/doc/ui.txt:
```diff
@@ -170,7 +174,9 @@ UIs).
 than that will be defined by future extensions. Just activating the `ext_newgrid`
-option by itself will never cause any additional grids to be created.
+option by itself will never cause any additional grids to be created. To
+enable per-window grids, `ext_multigrid` option must be set (see
```
To enable per-window grids, set the `ext_multigrid` option. (see |ui-multigrid|)

bfredl commented on src/nvim/api/ui.c:
```diff
@@ -268,12 +290,12 @@ static void remote_ui_grid_resize(UI *ui, Integer grid,
 {
   Array args = ARRAY_DICT_INIT;
-  if (ui->ui_ext[kUINewgrid]) {
+  if (ui->ui_ext[kUIMultigrid]) {
```
this change is not correct (already `ext_newgrid` uses `grid_resize`).

UtkarshMe replied:
Somehow terminal tests are failing after changing back to `kUINewgrid`. The problem is probably somewhere else. I'll see what I can find.

bfredl commented on src/nvim/mbyte.c:
```diff
@@ -558,9 +558,10 @@ size_t mb_string2cells(const char_u *str)
 /// Return number of display cells for char at ScreenLines[off].
 /// We make sure that the offset used is less than "max_off".
-int utf_off2cells(unsigned off, unsigned max_off)
+int utf_off2cells(ScreenGrid *grid, unsigned off, unsigned max_off)
```
this function can be deleted (screen.c uses now `line_off2cells`)

bfredl commented on src/nvim/screen.c:
```diff
@@ -504,7 +512,7 @@ void update_single_line(win_T *wp, linenr_T lnum)
 
   // Don't do anything if the screen structures are (not yet) valid.
-  if (!screen_valid(true) || updating_screen) {
+  if (linebuf_char == NULL || updating_screen) {
```
What is the reason for this change? Note `screen_valid` can also allocate the global grid in some situations if it wasn't already (though I don't know if it is still used).

bfredl commented on test/functional/ui/screen.lua:
```diff
@@ -181,33 +193,53 @@ function Screen:set_hlstate_cterm(val)
 function Screen:try_resize(columns, rows)
-  uimeths.try_resize(columns, rows)
+  self.uimeths.try_resize(columns, rows)
+  -- Give ourselves a chance to _handle_resize, which requires using
```
this sleep shouldn't be needed anymore.

bfredl commented on test/functional/ui/screen.lua:
```diff
@@ -181,33 +193,53 @@ function Screen:set_hlstate_cterm(val)
+function Screen:try_resize_grid(grid, columns, rows)
+  self._grid = self._grids[1]
+  self.uimeths.try_resize_grid(grid, columns, rows)
+  self:sleep(0.1)
```
neither this one.

bfredl commented on test/functional/ui/screen.lua:
```diff
@@ -181,33 +193,53 @@ function Screen:set_hlstate_cterm(val)
+end
+
+function Screen:try_resize_grid(grid, columns, rows)
+  self._grid = self._grids[1]
```
This is not needed here. Only legacy grid code uses it, and for it it is set in (legacy) `_handle_resize`.

bfredl commented on test/functional/ui/wildmode_spec.lua:
```diff
@@ -12,7 +12,7 @@ describe("'wildmenu'", function()
     clear()
     screen = Screen.new(25, 5)
-    screen:attach()
+    screen:attach({ext_newgrid=false})
```
Is this needed?

bfredl commented on test/helpers.lua:
```diff
@@ -48,8 +48,8 @@ local check_logs_useless_lines = {
 end
-local function neq(expected, actual)
-  return assert.are_not.same(expected, actual)
+local function neq(expected, actual, context)
```
Could change `eq` to also use `context` for consistency.

UtkarshMe commented on src/nvim/screen.c:
```diff
@@ -5394,18 +5394,18 @@ void grid_puts_len(ScreenGrid *grid, char_u *text, int textlen, int row,
 }
 
-void screen_puts_line_flush(bool set_cursor)
+void grid_puts_line_flush(ScreenGrid *grid, bool set_cursor)
```
Passing grid in `grid_puts_line_start` would mean creating another static global variable because we need `grid` in `grid_puts_line_flush` for `ui_line`. Or we can just pass the grid in both cases and assert that the grid is same?

breja commented:
How does this work with tabpages?

I'm trying to figure out how to draw windows associated to tabpages. For starters I'd like to hide any windows that do not belong to the current tabpage. I do not see an obvious method - currently `tabline_update` is sent after `win_position` events.

It would be neat if:
- `win_position` would also send the tab id that the window belongs to
  - e.g. `win_position(win: 1000, grid: 2, tab: 1, top: 0, left: 0, width: 20, height: 10)`
- some new event that relates a grid id to a tab id
- `tabline_update` sent *before* any `win_position` events - the client can then map all `win_position` events to the previous/last sent `tabline_update` current tab id
- an even better solution?

I have no idea about the complexity involved though...

bfredl commented:
Rather than asking the UI to keep track of which window belong to which tab, I think we should just send an `win_hide` event whenever a window becomes invisible due to a tab switched away. A `win_position` event should be sent whenever the window becomes visible again.

breja commented:
> ...send an `win_hide` event whenever a window becomes invisible due to a tab switched away

Yeah, I'm good with a hide event. Stateless UI is easier.

bfredl commented:
I hacked together a simple demo of multiple grid sizes here https://github.com/bfredl/python-gui/tree/winsize. Note: the python-gui code is quite messy as I basically reused some old code for floats and hacked in different inner/outer size with least possible effort, and shouldn't serve as "reference implementation" of any sort. However It might be useful for test driving and sorting out crashes and redraw glitches (many of which can be at client side, though)

Works:
 - changing current window font size with `:FontSize 12`

Doesn't work:
- tabs
- messages
- a lot of stuff that results in crashes (at either side)

![screenshot from 2018-10-16 16-33-22](https://user-images.githubusercontent.com/1363104/47024420-ef341180-d161-11e8-8d92-3a71fedd23e1.png)

bfredl commented:
I rebased the branch on recent conflicting changes I did (newgrid -> linegrid, various screen.lua refactors).

bfredl committed:
multigrid: send win_hide events when changing tabpage

bfredl commented:
Added `win_hide` event, sent whenever a window should become invisible due not to a visible tab anymore. `win_position` is sent when the window becomes visible again.

bfredl committed:
multigrid: add msg_scroll_start and msg_scroll_reset events

bfredl commented:
Added `msg_scroll_start` and `msg_scroll_reset` events. This works as a baseline to combine multigrid with multiline messages: after `scroll_start` is sent, scrolling on grid 1 should overdraw window grids, and when `scroll_reset` is sent, the overdrawn part of windows should be restored. #6619 uses this now to combine TUI floats and multiline messages correctly (as floats in TUI uses the multigrid protocol internally), if not with maximum efficiency.

Though the proper solution for non-`ext_messages` (#7466) clients (including TUI) is rather to draw long messages on a dedicated grid. This has benefits for TUI even if no floats are used: we can let messages.c render multi-line messages to the internal grid, and only send the final state to the terminal, unlike the current msgsep implementation (as well as the old vim implementation) which instead sends every message line directly to the terminal, and does a scroll for every line. This will especially make messages rendering a lot more [efficient](http://canonical.org/~kragen/eink-design) when using neovim on an [e-ink display](https://alternativebit.fr/posts/ultimate-writer/). That should probably be a followup PR however, but hopefully in the same release cycle (0.3.3).

breja commented:
I don't think there is anything stopping a remote UI from treating a standalone window as a floating window (I know I've thought about this). At that point, the differences between standalone windows and floating windows become less distinguishable. The remaining difference would be that a floating window can be anchored to another position on the grid. Anything else?

So I could see standalone windows and floating windows merged into one event (maybe even fixed rectangle windows) - that would make sense to me. Would the one window event then have various optional parameters?

Something like...
- standalone `win_position(win_id, grid_id, width, height)`
- fixed rectangle `win_position(win_id, grid_id, width, height, x, y)`
- floating `win_position(win_id, grid_id, width, height, x, y, anchor)`

Although at the end of the day, I'm okay with either one or three window positioning events. Whatever is easier on the Neovim end.

bfredl commented:
> I don't think there is anything stopping a remote UI from treating a standalone window as a floating window (I know I've thought about this). At that point, the differences between standalone windows and floating windows become less distinguishable. The remaining difference would be that a floating window can be anchored to another position on the grid. Anything else?

The UI can always do whatever, the point is events communicating reasonable expectations. The difference is that an anchored window has its position managed by a plugin or something (because the window provides information to some text in a window), while a standalone window would have its position managed by the UI or the user indirectly. Rendering a standalone window inline the same pixbuf but with some kind of moving handle for the user would also be a valid interpretation. 

The problem with the "optional arg" approach is that the difference is not that "more information is available" vs "the UI needs to fill it in" which it would seem to mean at a first glance, but rather fixed-rect vs anchor has different semantics even for the shared args: for a fixed-rect the size is the total size the window should occupy on the main grid, including scroll bars or whatever widgets the UI wants to add on the border. While for an anchored float, the only size that is part of the protocol is the grid size of the float itself (which even makes width/height args redundant, as `grid_resize`events always is sent before).

bfredl committed:
multigrid: rename event to win_pos, make grid first

bfredl committed:
multigrid: test refactor

bfredl committed:
multigrid: add tests for multiline messages and tabs

bfredl commented:
Added docs and tests for the support of multiline-messages and tabs.

coditva committed:
multigrid: add multigrid screen tests

bfredl committed:
multigrid: doc update

bfredl committed:
multigrid: various cleanup (types, unused parameters)

Handle the rare case of full highlight table properly

bfredl committed:
multigrid: reorganize types and global varaibles

bfredl committed:
multigrid: rename to grid.row_offset and grid.requested_rows

bfredl committed:
multigrid: simplify tests

bfredl committed:
multigrid: API version bump

bfredl committed:
multigrid: rename grid->ScreenLines and other grid arrays

bfredl commented:
Merged after some further cleanup and doc changes. Thanks @UtkarshMe for this significant work that will enable many interesting enhancements to Nvim GUIs.

Dealing with long issues/PR threads in github is quite unwieldy. I'll open a new issue where follow-up work and issues can be discussed.
